File: OpenBurnBarCore/Sources/OpenBurnBarCore/Hermes/HermesAtom.swift
import Foundation

// MARK: - Hermes Atom
//
// Typed enum of "conversation atoms" — entities that Hermes references in
// chat which the OpenBurnBar app already has dedicated UI for. Atoms are
// emitted by Hermes (or detected from prose) and rendered as atomic
// inline chips that, when tapped, navigate to the matching native view.
//
// Pretext rich-inline ensures these chips never break across lines and
// that prose flows naturally around them.

/// Time window selector reused across the app's analytics surfaces.
public enum HermesAtomWindow: String, Hashable, Codable, Sendable, CaseIterable {
    case today
    case yesterday
    case sevenDays = "7d"
    case thirtyDays = "30d"
    case ninetyDays = "90d"
    case all

    public var displayLabel: String {
        switch self {
        case .today:      return "today"
        case .yesterday:  return "yesterday"
        case .sevenDays:  return "7 days"
        case .thirtyDays: return "30 days"
        case .ninetyDays: return "90 days"
        case .all:        return "all time"
        }
    }
}

/// Categorical group of token-volume scopes.
public enum HermesAtomTokenScope: String, Hashable, Codable, Sendable {
    case today
    case session
    case run
    case lifetime
    case unspecified

    public var displayLabel: String {
        switch self {
        case .today:       return "today"
        case .session:     return "this session"
        case .run:         return "this run"
        case .lifetime:    return "lifetime"
        case .unspecified: return ""
        }
    }
}

/// One conversation atom — a strongly-typed reference to an entity in the
/// app. Each case carries enough data to navigate to the entity, plus a
/// `HermesAtomKind` that drives icon / accent / label format choices.
public enum HermesAtom: Hashable, Sendable {
    /// A monetary cost across some time window. Decimal-as-Double avoids
    /// the Codable headache of bridging Decimal through a JSON-able
    /// serialization layer; cost atoms are display-only.
    case cost(amount: Double, window: HermesAtomWindow)
    /// A specific session, identified by its persistent ID.
    case session(id: String)
    /// A provider (e.g. anthropic, openai, kimi). Token matches
    /// `AgentProvider.fromPersistedToken` in the app.
    case provider(token: String)
    /// A specific model, by ID (e.g. `claude-sonnet-4.7`).
    case model(id: String)
    /// A time window selector (e.g. tap to switch dashboard to 7d).
    case window(HermesAtomWindow)
    /// A tool the assistant called.
    case tool(name: String)
    /// A project tracked in OpenBurnBar.
    case project(id: String)
    /// A token volume — counts that the app already surfaces in headers
    /// and detail sheets.
    case tokens(value: Int, scope: HermesAtomTokenScope)
    /// Quota for a specific provider, expressed as a percentage used.
    case quota(provider: String, percent: Int)
    /// A Hermes runtime profile (e.g. `hermes`, `relay`, `local`).
    case runtime(profile: String)

    public var kind: HermesAtomKind {
        switch self {
        case .cost:     return .cost
        case .session:  return .session
        case .provider: return .provider
        case .model:    return .model
        case .window:   return .window
        case .tool:     return .tool
        case .project:  return .project
        case .tokens:   return .tokens
        case .quota:    return .quota
        case .runtime:  return .runtime
        }
    }

    /// Default label used when the source label is missing or whitespace-
    /// only. Keeps the rendered chip from collapsing.
    public var fallbackLabel: String {
        switch self {
        case .cost(let amount, let window):
            let formatter = NumberFormatter()
            formatter.numberStyle = .currency
            formatter.currencyCode = "USD"
            formatter.maximumFractionDigits = 2
            let cost = formatter.string(from: NSNumber(value: amount)) ?? "$\(amount)"
            return "\(cost) \(window.displayLabel)"
        case .session(let id):     return "session \(id.prefix(8))"
        case .provider(let token): return token.capitalized
        case .model(let id):       return id
        case .window(let w):       return w.displayLabel
        case .tool(let name):      return name
        case .project(let id):     return id
        case .tokens(let value, let scope):
            let formatted = HermesAtom.formatTokenCount(value)
            return scope == .unspecified ? "\(formatted) tokens" : "\(formatted) \(scope.displayLabel)"
        case .quota(let provider, let percent):
            return "\(percent)% \(provider.capitalized)"
        case .runtime(let profile):
            return profile.capitalized
        }
    }

    /// Human-friendly token count formatter — `12.4k`, `1.2M`, etc.
    public static func formatTokenCount(_ value: Int) -> String {
        if value < 1_000 { return "\(value)" }
        if value < 1_000_000 {
            let k = Double(value) / 1_000
            return String(format: "%.1fk", k)
        }
        let m = Double(value) / 1_000_000
        return String(format: "%.1fM", m)
    }
}

// MARK: - Atom Kind (presentation metadata)

/// Visual / behavioral category for an atom. Maps to icon, accent color
/// token name, and short display category. The actual SF Symbol + Color
/// resolution lives in each app's UI layer (different design systems for
/// iOS vs macOS).
public enum HermesAtomKind: String, Hashable, Codable, Sendable, CaseIterable {
    case cost
    case session
    case provider
    case model
    case window
    case tool
    case project
    case tokens
    case quota
    case runtime

    /// SF Symbol identifier suitable for the chip's leading glyph.
    public var systemImage: String {
        switch self {
        case .cost:     return "dollarsign.circle.fill"
        case .session:  return "rectangle.stack.fill"
        case .provider: return "externaldrive.connected.to.line.below"
        case .model:    return "cpu"
        case .window:   return "calendar"
        case .tool:     return "wrench.and.screwdriver.fill"
        case .project:  return "folder.fill"
        case .tokens:   return "number"
        case .quota:    return "gauge.with.dots.needle.67percent"
        case .runtime:  return "antenna.radiowaves.left.and.right"
        }
    }

    /// Short category label used by the detail sheet header.
    public var categoryLabel: String {
        switch self {
        case .cost:     return "Cost"
        case .session:  return "Session"
        case .provider: return "Provider"
        case .model:    return "Model"
        case .window:   return "Window"
        case .tool:     return "Tool"
        case .project:  return "Project"
        case .tokens:   return "Tokens"
        case .quota:    return "Quota"
        case .runtime:  return "Runtime"
        }
    }

    /// One-line explanation used inside the detail sheet.
    public var description: String {
        switch self {
        case .cost:     return "Open the burn detail for this time window."
        case .session:  return "Open this session's detail view."
        case .provider: return "Open this provider's dashboard."
        case .model:    return "Open this model's detail or pick it as default."
        case .window:   return "Switch the dashboard to this time window."
        case .tool:     return "See where this tool was invoked in the run."
        case .project:  return "Open this project's detail."
        case .tokens:   return "Open the token-usage detail."
        case .quota:    return "Open quota detail for this provider."
        case .runtime:  return "Open Hermes runtime details for this profile."
        }
    }
}

File: windows/app/OpenBurnBar.App.Presentation/Chat/HermesAtom.cs
using System;
using System.Globalization;

namespace OpenBurnBar.App.Presentation.Chat;

// MARK: - Hermes Atom
//
// C# peer of `OpenBurnBarCore/Sources/OpenBurnBarCore/Hermes/HermesAtom.swift`.
//
// Typed set of "conversation atoms" — entities Hermes references in chat which
// the OpenBurnBar app already has dedicated UI for. Atoms are emitted by Hermes
// (or detected from prose) and rendered as atomic inline chips that, when
// tapped, navigate to the matching native surface.
//
// Swift models this as an `enum HermesAtom` with associated values. C# has no
// enum-with-payload, so each case becomes a small `sealed record` under the
// abstract `HermesAtom` base — records give the same value-equality +
// hashability the Swift `Hashable` conformance relied on, and `switch`
// expressions over the record types recover exhaustive pattern matching.

/// Time window selector reused across the app's analytics surfaces
/// (peer of Swift `HermesAtomWindow`). Raw values match the Swift `rawValue`
/// so `burnbar://` URLs round-trip byte-for-byte.
public enum HermesAtomWindow
{
    Today,
    Yesterday,
    SevenDays,
    ThirtyDays,
    NinetyDays,
    All,
}

/// Categorical group of token-volume scopes (peer of Swift `HermesAtomTokenScope`).
public enum HermesAtomTokenScope
{
    Today,
    Session,
    Run,
    Lifetime,
    Unspecified,
}

/// Visual / behavioral category for an atom (peer of Swift `HermesAtomKind`).
/// Drives icon / accent / label choices; the concrete glyph + color resolution
/// lives in the WinUI layer (a different design system than macOS SF Symbols).
public enum HermesAtomKind
{
    Cost,
    Session,
    Provider,
    Model,
    Window,
    Tool,
    Project,
    Tokens,
    Quota,
    Runtime,
}

/// Raw-value helpers so the URL codec and window/scope labels match Swift's
/// `String` raw values exactly.
public static class HermesAtomWindowRaw
{
    public static string ToRawValue(this HermesAtomWindow window) => window switch
    {
        HermesAtomWindow.Today => "today",
        HermesAtomWindow.Yesterday => "yesterday",
        HermesAtomWindow.SevenDays => "7d",
        HermesAtomWindow.ThirtyDays => "30d",
        HermesAtomWindow.NinetyDays => "90d",
        HermesAtomWindow.All => "all",
        _ => "today",
    };

    public static HermesAtomWindow? FromRawValue(string? raw) => raw switch
    {
        "today" => HermesAtomWindow.Today,
        "yesterday" => HermesAtomWindow.Yesterday,
        "7d" => HermesAtomWindow.SevenDays,
        "30d" => HermesAtomWindow.ThirtyDays,
        "90d" => HermesAtomWindow.NinetyDays,
        "all" => HermesAtomWindow.All,
        _ => null,
    };

    public static string DisplayLabel(this HermesAtomWindow window) => window switch
    {
        HermesAtomWindow.Today => "today",
        HermesAtomWindow.Yesterday => "yesterday",
        HermesAtomWindow.SevenDays => "7 days",
        HermesAtomWindow.ThirtyDays => "30 days",
        HermesAtomWindow.NinetyDays => "90 days",
        HermesAtomWindow.All => "all time",
        _ => "today",
    };
}

public static class HermesAtomTokenScopeRaw
{
    public static string ToRawValue(this HermesAtomTokenScope scope) => scope switch
    {
        HermesAtomTokenScope.Today => "today",
        HermesAtomTokenScope.Session => "session",
        HermesAtomTokenScope.Run => "run",
        HermesAtomTokenScope.Lifetime => "lifetime",
        HermesAtomTokenScope.Unspecified => "unspecified",
        _ => "unspecified",
    };

    public static HermesAtomTokenScope? FromRawValue(string? raw) => raw switch
    {
        "today" => HermesAtomTokenScope.Today,
        "session" => HermesAtomTokenScope.Session,
        "run" => HermesAtomTokenScope.Run,
        "lifetime" => HermesAtomTokenScope.Lifetime,
        "unspecified" => HermesAtomTokenScope.Unspecified,
        _ => null,
    };

    public static string DisplayLabel(this HermesAtomTokenScope scope) => scope switch
    {
        HermesAtomTokenScope.Today => "today",
        HermesAtomTokenScope.Session => "this session",
        HermesAtomTokenScope.Run => "this run",
        HermesAtomTokenScope.Lifetime => "lifetime",
        HermesAtomTokenScope.Unspecified => "",
        _ => "",
    };
}

/// One conversation atom — a strongly-typed reference to an app entity, carrying
/// enough data to navigate plus a <see cref="Kind"/> that drives icon/accent/label.
public abstract record HermesAtom
{
    private HermesAtom()
    {
    }

    public abstract HermesAtomKind Kind { get; }

    /// A monetary cost across some time window (display-only; Decimal-as-double
    /// mirrors the Swift note about avoiding Decimal Codable bridging).
    public sealed record Cost(double Amount, HermesAtomWindow Window) : HermesAtom
    {
        public override HermesAtomKind Kind => HermesAtomKind.Cost;
    }

    public sealed record Session(string Id) : HermesAtom
    {
        public override HermesAtomKind Kind => HermesAtomKind.Session;
    }

    // Named `ProviderRef` (not `Provider`) so it does not collide with the
    // positional parameter `Provider` on the `Quota` record below (C# resolves
    // the nested type over the synthesized property otherwise → CS8866).
    public sealed record ProviderRef(string Token) : HermesAtom
    {
        public override HermesAtomKind Kind => HermesAtomKind.Provider;
    }

    public sealed record Model(string Id) : HermesAtom
    {
        public override HermesAtomKind Kind => HermesAtomKind.Model;
    }

    // Named `WindowRef` (not `Window`) so it does not collide with the positional
    // parameter `Window` on the `Cost` record above (same CS8866 reason).
    public sealed record WindowRef(HermesAtomWindow Value) : HermesAtom
    {
        public override HermesAtomKind Kind => HermesAtomKind.Window;
    }

    public sealed record Tool(string Name) : HermesAtom
    {
        public override HermesAtomKind Kind => HermesAtomKind.Tool;
    }

    public sealed record Project(string Id) : HermesAtom
    {
        public override HermesAtomKind Kind => HermesAtomKind.Project;
    }

    public sealed record Tokens(int Value, HermesAtomTokenScope Scope) : HermesAtom
    {
        public override HermesAtomKind Kind => HermesAtomKind.Tokens;
    }

    public sealed record Quota(string Provider, int Percent) : HermesAtom
    {
        public override HermesAtomKind Kind => HermesAtomKind.Quota;
    }

    public sealed record Runtime(string Profile) : HermesAtom
    {
        public override HermesAtomKind Kind => HermesAtomKind.Runtime;
    }

    /// Default label used when the source label is missing or whitespace-only.
    /// Byte-for-byte port of Swift `HermesAtom.fallbackLabel`.
    public string FallbackLabel => this switch
    {
        Cost c => $"{FormatCurrency(c.Amount)} {c.Window.DisplayLabel()}",
        Session s => $"session {Prefix(s.Id, 8)}",
        ProviderRef p => Capitalized(p.Token),
        Model m => m.Id,
        WindowRef w => w.Value.DisplayLabel(),
        Tool t => t.Name,
        Project pr => pr.Id,
        Tokens tk => tk.Scope == HermesAtomTokenScope.Unspecified
            ? $"{FormatTokenCount(tk.Value)} tokens"
            : $"{FormatTokenCount(tk.Value)} {tk.Scope.DisplayLabel()}",
        Quota q => $"{q.Percent}% {Capitalized(q.Provider)}",
        Runtime r => Capitalized(r.Profile),
        _ => "",
    };

    /// Human-friendly token count formatter — `12.4k`, `1.2M`, `3.40B`.
    /// Byte-for-byte port of Swift `HermesAtom.formatTokenCount`.
    public static string FormatTokenCount(int value)
    {
        if (value < 1_000)
        {
            return value.ToString(CultureInfo.InvariantCulture);
        }
        if (value < 1_000_000)
        {
            var k = value / 1_000.0;
            return string.Format(CultureInfo.InvariantCulture, "{0:0.0}k", k);
        }
        if (value >= 1_000_000_000)
        {
            var b = value / 1_000_000_000.0;
            return string.Format(CultureInfo.InvariantCulture, "{0:0.00}B", b);
        }
        var m = value / 1_000_000.0;
        return string.Format(CultureInfo.InvariantCulture, "{0:0.0}M", m);
    }

    // Mirror of Swift's `NumberFormatter(.currency, USD, maxFractionDigits: 2)`
    // for the common en-US formatting the chip renders.
    private static string FormatCurrency(double amount) =>
        amount.ToString("C2", CultureInfo.GetCultureInfo("en-US"));

    private static string Capitalized(string value)
    {
        if (string.IsNullOrEmpty(value))
        {
            return value;
        }
        return char.ToUpper(value[0], CultureInfo.InvariantCulture) + value.Substring(1);
    }

    private static string Prefix(string value, int count) =>
        value.Length <= count ? value : value.Substring(0, count);
}

/// Presentation metadata for a <see cref="HermesAtomKind"/> — category label +
/// one-line description. The WinUI layer maps <see cref="HermesAtomKind"/> to a
/// Segoe glyph + accent brush (peer of the Swift `systemImage`/`categoryLabel`/
/// `description` computed properties, minus the Apple-only SF Symbol names).
public static class HermesAtomKindMetadata
{
    public static string CategoryLabel(this HermesAtomKind kind) => kind switch
    {
        HermesAtomKind.Cost => "Cost",
        HermesAtomKind.Session => "Session",
        HermesAtomKind.Provider => "Provider",
        HermesAtomKind.Model => "Model",
        HermesAtomKind.Window => "Window",
        HermesAtomKind.Tool => "Tool",
        HermesAtomKind.Project => "Project",
        HermesAtomKind.Tokens => "Tokens",
        HermesAtomKind.Quota => "Quota",
        HermesAtomKind.Runtime => "Runtime",
        _ => "",
    };

    public static string Description(this HermesAtomKind kind) => kind switch
    {
        HermesAtomKind.Cost => "Open the burn detail for this time window.",
        HermesAtomKind.Session => "Open this session's detail view.",
        HermesAtomKind.Provider => "Open this provider's dashboard.",
        HermesAtomKind.Model => "Open this model's detail or pick it as default.",
        HermesAtomKind.Window => "Switch the dashboard to this time window.",
        HermesAtomKind.Tool => "See where this tool was invoked in the run.",
        HermesAtomKind.Project => "Open this project's detail.",
        HermesAtomKind.Tokens => "Open the token-usage detail.",
        HermesAtomKind.Quota => "Open quota detail for this provider.",
        HermesAtomKind.Runtime => "Open Hermes runtime details for this profile.",
        _ => "",
    };
}
